broker/src/bin/neon_broker.rs

```diff
@@ -0,0 +1,563 @@
+            (id, pub_txs, announce) = ss.register_subscriber(sub_key, tx);
+        }
+        // Note: it is important to create Subscriber before .await. If client
+        // disconnects during await, which would terminate the Future we still
+        // need to run Subscriber's drop() which will unregister it from the
+        // shared state.
+        let subscriber = Subscriber {
+            id,
+            key: sub_key,
+            sub_rx: rx,
+            registry: self.clone(),
+        };
+        // Notify existing publishers about new subscriber.
+        for pub_tx in pub_txs.iter_mut() {
+            // Closed channel is fine; it means publisher has gone.
+            pub_tx.send(announce.clone()).await.ok();
+        }
+        trace!("registered subscriber {}", id);
+        subscriber
+    }
+
+    // Unregister the subscriber
+    pub fn unregister_subscriber(&self, sub: &Subscriber) {
+        let mut ss = self.shared_state.lock().unwrap();
+        let announce_pack = ss.unregister_subscriber(sub.id, sub.key);
+        drop(ss);
+        // Notify publishers about the removal. Apart from wanting to do it
+        // outside lock, here we also spin a task as Drop impl can't be async.
+        if let Some((mut pub_txs, announce)) = announce_pack {
+            tokio::spawn(async move {
+                for pub_tx in pub_txs.iter_mut() {
+                    // Closed channel is fine; it means publisher has gone.
+                    pub_tx.send(announce.clone()).await.ok();
+                }
+            });
+        }
+        trace!("unregistered subscriber {}", sub.id);
+    }
+}
+
+// Private subscriber state.
+struct Subscriber {
+    id: SubId,
+    key: SubscriptionKey,
+    // Subscriber receives messages from publishers here.
+    sub_rx: Receiver<SafekeeperTimelineInfo>,
+    // to unregister itself from shared state in Drop
+    registry: Registry,
+}
+
+impl Drop for Subscriber {
+    fn drop(&mut self) {
+        self.registry.unregister_subscriber(self);
+    }
+}
+
+// Private publisher state
+struct Publisher {
+    id: PubId,
+    // new subscribers request to send (or stop sending) msgs them here.
+    // It could be just Receiver, but weirdly it doesn't implement futures_core Stream directly.
+    announce_rx: ReceiverStream<SubAnnounce>,
+    subs_to_all: Vec<Sender<SafekeeperTimelineInfo>>,
+    subs_to_timelines: HashMap<TenantTimelineId, Vec<SubSender>>,
+    // to unregister itself from shared state in Drop
+    registry: Registry,
+}
+
+impl Publisher {
+    // Send msg to relevant subscribers.
+    pub fn send_msg(&mut self, msg: &SafekeeperTimelineInfo) -> Result<(), Status> {
+        // send message to subscribers for everything
+        let mut cleanup_subs_to_all = false;
+        for sub in self.subs_to_all.iter() {
+            match sub.try_send(msg.clone()) {
+                Err(TrySendError::Full(_)) => {
+                    trace!("dropping message, channel is full");
```

Comment:
I think we did not change that and the other `trace!` to warn (or added the metrics, or something else)
See https://github.com/neondatabase/neon/pull/2733#discussion_r1012156020